lib/Runtime/ByteCode/ByteCodeEmitter.cpp

```diff
@@ -5572,28 +5572,65 @@ void ByteCodeGenerator::EmitPropTypeof(Js::RegSlot lhsLocation, Symbol *sym, Ide
 
 void ByteCodeGenerator::EnsureNoRedeclarations(ParseNodeBlock *pnodeBlock, FuncInfo *funcInfo)
 {
-    // Emit dynamic runtime checks for variable re-declarations. Only necessary for global functions (script or eval).
+    // Emit dynamic runtime checks for re-declarations. Only necessary for global functions (script or eval).
     // In eval only var declarations can cause redeclaration, and only in non-strict mode, because let/const variables
-    // remain local to the eval code.
+    // and functions remain local to the eval code.
 
     Assert(pnodeBlock->nop == knopBlock);
     Assert(pnodeBlock->blockType == PnodeBlockType::Global || pnodeBlock->scope->GetScopeType() == ScopeType_GlobalEvalBlock);
 
+    auto emitEnsureNoRootFld = [this](Symbol * sym, Js::OpCode opCode)
+    {
+        FuncInfo *funcInfo = this->TopFuncInfo();
+        Assert(sym->GetIsGlobal());
+        Js::PropertyId propertyId = sym->EnsurePosition(this);
+        this->m_writer.ElementRootU(opCode, funcInfo->FindOrAddReferencedPropertyId(propertyId));
+    };
+
     if (!(this->flags & fscrEvalCode))
     {
-        IterateBlockScopedVariables(pnodeBlock, [this](ParseNode *pnode)
+        // scan for let/const
+        // these will conflict with all global properties including other global let and const variables
+        IterateBlockScopedVariables(pnodeBlock, [&](ParseNode *pnode)
         {
-            FuncInfo *funcInfo = this->TopFuncInfo();
-            Symbol *sym = pnode->AsParseNodeVar()->sym;
+            emitEnsureNoRootFld(pnode->AsParseNodeVar()->sym, Js::OpCode::EnsureNoRootFld);
+        });
+    }
 
-            Assert(sym->GetIsGlobal());
+    // scan for function declarations
```

Comment:
Should we be scanning for block-scoped functions in strict mode? Don't they shadow same-named globals rather than conflicting with them?

Reply:
yes, in strict mode they should just shadow. 


---
In reply to: [184167130](https://github.com/Microsoft/ChakraCore/pull/5042#discussion_r184167130) [](ancestors = 184167130)

Reply:
I also wonder if nesting in if/for blocks matters, like:
```js

// in a loaded/eval'd  script

if (true)
{
    function foo(){};
}
```

---
In reply to: [184217835](https://github.com/Microsoft/ChakraCore/pull/5042#discussion_r184217835) [](ancestors = 184217835,184167130)

Reply:
strict mode is handled now the same way as for `var` in the same situation

---
In reply to: [184218209](https://github.com/Microsoft/ChakraCore/pull/5042#discussion_r184218209) [](ancestors = 184218209,184217835,184167130)